test/framework/src/main/java/org/elasticsearch/test/RandomObjects.java

```diff
@@ -185,14 +185,21 @@ public static BytesReference randomSource(Random random, XContentType xContentTy
 
     /**
      * Randomly adds fields, objects, or arrays to the provided builder. The maximum depth is 5.
+     * @return Returns the number of fields in the document
      */
-    private static void addFields(Random random, XContentBuilder builder, int minNumFields, int currentDepth) throws IOException {
-        int numFields = randomIntBetween(random, minNumFields, 10);
+    private static int addFields(Random random, XContentBuilder builder, int minNumFields, int currentDepth,
+                                 int currentFields) throws IOException {
+        int maxTotalFields = 200;
+        int maxFields = Math.max(minNumFields, 10 - (currentFields * 10)/maxTotalFields); // Map to range 0-10
+        int numFields = randomIntBetween(random, minNumFields, maxFields);
+        int maxDepth = 5 - (currentFields * 5)/maxTotalFields; // Map to range 0-5
+        currentFields += numFields;
```

Comment:
with the goal of trying to keep things simple, I wonder if decreasing the max number of fields e.g. to 5 would be enough. What do you think?

Reply:
Could be. Will need to test this for the average case. The potentially maximum number of fields would still be more than 3000.

Reply:
maybe we should also decrease the chance that we add another child object compared to leaf fields.

Reply:
I didn't quite understand. We, already have a check on the depth which takes care of this? Or do you mean something else?

Reply:
yes we have a check, but I think we end up adding more objects and arrays than we need.

```
if (currentDepth < 5 && random.nextBoolean()) {
```

the above if (taken from addFields method) means that we add a new object or array if maxDepth allows, one out of two times. Maybe we want to lower that probability?

Reply:
Ah okay. I will try to lower the probability based on current depth.

Reply:
you can also do something similar to what LuceneTestCase#rarely does. we can do it regardless depth to keep things simple?

Reply:
Yes. Okay.

Reply:
@javanna I simplified it now. So that max fields is 5 and the probability is 30%. It seems to give similar results in terms of timing.